### PR TITLE
doc(frameworks): tensorflow

### PR DESCRIPTION
**Pull request by bojiang**
## Checklist:

- [ ] My code follows the bentoml code style, both `make format` and
  `make lint` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: docs/source/frameworks/tensorflow.rst [hunk 1]
```diff
@@ -2,51 +2,236 @@
...
+                ]
+            )
+            def __call__(self, inputs):
+                return self.dense(inputs)
+
+
+        model = NativeModel()
+        ... # training
+
+        # =========
```
Comment (ssheng) [inline docs/source/frameworks/tensorflow.rst:61]: Can we add some bare minimum training logic so user can run the example.

File: docs/source/frameworks/tensorflow.rst [hunk 2]
```diff
@@ -2,51 +2,236 @@
...
+   `bentoml.tensorflow.save`_ has two parameters `tf_signature` and `signatures`.
+   They are important when you want to save a model with a ensured behavior.
+   The `tf_signature` is a dict of tensor names and their shapes, which adaopted from
+   the signature of tensorflow.saved_model.save. You may find more details about it
+   on the `tensorflow.saved_model.save`_ documentation.
+   The `signatures` is a dict of functions names and some other information.
+   If you know your model has a dynamic batch dimension, you can use `signatures` to tell
+   bentoml about that for possible future optimizations like this:
+
+   bentoml.tensorflow.save(model, "my_model", signatures={"__call__": {"batch_dim": 0, "batchable": True}})
```
Comment (ssheng) [inline docs/source/frameworks/tensorflow.rst:133]: Should `tf_signature` and `signatures` arguments be mutually exclusive? Let's also link the TF documentation here.

File: docs/source/frameworks/tensorflow.rst [hunk 3]
```diff
@@ -2,51 +2,236 @@
...
+            ...
+
+    model = MultiInputModel()
+    ... # training
+
+    bentoml.tensorflow.save(model, "my_tf_model")
+
+.. seealso::
+   `bentoml.tensorflow.save`_ has two parameters `tf_signature` and `signatures`.
+   They are important when you want to save a model with a ensured behavior.
```
Comment (ssheng) [inline docs/source/frameworks/tensorflow.rst:125]: > They are important when you want to save a model with a ensured behavior.

Could you please elaborate on what this means?

File: docs/source/frameworks/tensorflow.rst [hunk 4]
```diff
@@ -2,51 +2,236 @@
...
+------------------------------
+
+.. code-block:: python
+
+    runner = bentoml.tensorflow.get("my_tf_model").to_runner()
+
+    runner.init_local()  # only for testing, do not call this in a bento service definition
+    runner.__call__.run(input_data)
+    # the same as:
+    # runner.run(input_data)
```
Comment (ssheng) [inline docs/source/frameworks/tensorflow.rst:146]: I think this step is unnecessary as it is neither needed to build a service nor specific to TensorFlow.

File: docs/source/frameworks/tensorflow.rst [hunk 5]
```diff
@@ -2,51 +2,236 @@
...
+    #client 1
+    runner.run([[1,2,3,4,5]])
 
-.. autofunction:: bentoml.tensorflow.load_model
+    #client 2
+    runner.run([[6,7,8,9,0]])
 
-.. autofunction:: bentoml.tensorflow.get
+    # if multiple requests from different clients arrived at the same time,
+    # bentoml will automatically merge them and call model([[1,2,3,4,5], [6,7,8,9,0]])
```
Comment (ssheng) [inline docs/source/frameworks/tensorflow.rst:228]: This section may not be needed if we link the adaptive batching documentation.

File: docs/source/frameworks/tensorflow.rst [hunk 6]
```diff
@@ -1,52 +1,244 @@
...
-    import tensorflow as tf
+Consider using BentoML :ref:`frameworks/keras:Keras` framework if working with a Keras model. If continuing with TensorFlow, make sure the Keras model 
+is decorated with :code:`tf.function`.
 
-    class NativeModel(tf.Module):
+.. note::
+
+    - Keras is not optimized for inference in production. There are reports about memory leaks during serving. :code:`bentoml.keras` has same concerns since it rely on the the Keras APIs.
+    - Inference with :code:`bentoml.tensorflow` is about twice faster than :code:`bentoml.keras`.
+    - :code:`bentoml.keras` performs input casting similar to the original Keras model for better debugging experiences.
```
Comment (ssheng) [inline docs/source/frameworks/tensorflow.rst:18]: Should this be a part of the Keras guide instead?

File: docs/source/frameworks/tensorflow.rst [hunk 7]
```diff
@@ -1,52 +1,244 @@
...
+
+.. note::
+
+    :code:`bentoml.tensorflow.save_model` has two parameters :code:`tf_signature` and :code:`signatures`.
+    They are important when you want to save a model with a ensured behavior.
+    :code:`tf_signature`, inspired by :code:`tensorflow.saved_model.save`, is a dict of tensor names and their shapes. 
+    You may find more details about it on the :code:`tensorflow.saved_model.save` documentation.
+    The :code:`signatures` is a dict of functions names and some other information.
+    If you know your model has a dynamic batch dimension, you can use `signatures` to tell
+    bentoml about that for possible future optimizations like this:
```
Comment (aarnphm) [inline docs/source/frameworks/tensorflow.rst:132]:
```suggestion
    To ensure the Tensorflow model signature are consistent between loading in a Python session and loading from the BentoML model store, the implementation are as follow:
    - `tf_signature` is an alias to `tf.saved_model.save <https://www.tensorflow.org/api_docs/python/tf/saved_model/save>`_ *signatures* field. This optional signatures controls which methods in a given `obj <https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/trackable/base.py#L281>`_ will be available to programs that consume `SavedModel's <https://www.tensorflow.org/guide/saved_model>`_, for example, serving APIs. Read more about Tensorflow's signatures behaviour `from their API documentation <https://www.tensorflow.org/api_docs/python/tf/saved_model/save>`_.
    - ``signatures`` refers to a general :ref:`Model Signatures <concepts/model:Model Signatures>`_ that dictates which methods can be used for inference in the Runner context. This signature dictionary will be used during the creation process of a Runner instance.

The default signature for creating a Runner is ``{"__call__": {"batchable": False}}``. The means by default, BentoML’s `Adaptive Batching <guides/batching:Adaptive Batching>`_ is disabled when using :obj:`~bentoml.tensorflow.save_model()`. If you want to utilize adaptive batching behaviour and know your model's dynamic batching dimension, make sure to pass in ``signatures`` as follow: 

```

**Review by aarnphm — COMMENTED** [review]
Preliminary thoughts and comments. You probably have to pull down and merge some format changes.

File: docs/source/frameworks/tensorflow.rst [hunk 8]
```diff
@@ -1,52 +1,244 @@
 ==========
-Tensorflow
+TensorFlow
 ==========
 
-Here's an example of serving Tensorflow models with BentoML:
+TensorFlow is an open source machine learning library focusing on training on inference of deep neural networks. BentoML provides native support for 
+serving and deploying models trained from TensorFlow.
```
Comment (sauyon) [inline docs/source/frameworks/tensorflow.rst:6]:
```suggestion
TensorFlow is an open source machine learning library focusing on deep neural networks. BentoML provides native support for 
serving and deploying models trained from TensorFlow.
```

File: docs/source/frameworks/tensorflow.rst [hunk 9]
```diff
@@ -1,52 +1,244 @@
...
 
-    import bentoml
-    import tensorflow as tf
+Consider using BentoML :ref:`frameworks/keras:Keras` framework if working with a Keras model. If continuing with TensorFlow, make sure the Keras model 
+is decorated with :code:`tf.function`.
 
-    class NativeModel(tf.Module):
+.. note::
+
+    - Keras is not optimized for inference in production. There are reports about memory leaks during serving. :code:`bentoml.keras` has same concerns since it rely on the the Keras APIs.
```
[the remark this answers is not in the document]
Comment (sauyon) [reply docs/source/frameworks/tensorflow.rst:16]: Can remove the next bullet point this way, I think.

```suggestion
    - Keras is not optimized for inference in production. There are `many reports <https://github.com/tensorflow/tensorflow/issues?q=is%3Aissue+sort%3Aupdated-desc+keras+memory+leak>`_ of memory leaks during serving. Use :obj:`~bentoml.tensorflow`, as it is about twice as fast.
```

File: docs/source/frameworks/tensorflow.rst (hunk 6 above)
Comment (sauyon) [reply docs/source/frameworks/tensorflow.rst:18]: Do / should we have a separate Keras guide?

File: docs/source/frameworks/tensorflow.rst [hunk 10]
```diff
@@ -1,52 +1,244 @@
...
+    - Inference with :code:`bentoml.tensorflow` is about twice faster than :code:`bentoml.keras`.
+    - :code:`bentoml.keras` performs input casting similar to the original Keras model for better debugging experiences.
+
+To improve the performance of Keras models, consider applying techniques like **model distillation** or **model quantization**. Alternatively, the Keras model can be converted to a ONNX model and saved with :code:`bentoml.onnx` to leverage better performance runtimes (for eg: TensorRT).
```
[the remark this answers is not in the document]
Comment (sauyon) [reply docs/source/frameworks/tensorflow.rst:20]:
```suggestion
.. note::

    :bdg-info:`Remarks:` We recommend users apply model optimization techniques such as **distillation** or **quantization**. Alternatively, Keras models can also be converted to :ref:`ONNX <frameworks/onnx>` models and leverage different runtimes (e.g. TensorRT, Apache TVM, etc.).
```

File: docs/source/frameworks/tensorflow.rst [hunk 11]
```diff
@@ -1,52 +1,244 @@
...
+Compatibility
+-------------
+
+BentoML requires TensorFlow version 2.0 or higher. For TensorFlow version 1.0, consider using a :ref:`concepts/runner:Custom Runner`.
+
+
+Saving a Trained Model
+----------------------
+
+`bentoml.tensorflow` supports saving models in the formats of :code:`tf.Module`, :code:`keras.models.Sequential`, or :code:`keras.Model`.
```
Comment (sauyon) [inline docs/source/frameworks/tensorflow.rst:31]:
```suggestion
``bentoml.tensorflow`` supports saving ``tf.Module``s, ``keras.models.Sequential``s, and ``keras.Model``s.
```

File: docs/source/frameworks/tensorflow.rst [hunk 12]
```diff
@@ -1,52 +1,244 @@
...
+        y = keras.layers.Dense(
+            6,
+            name="out",
+            kernel_initializer=keras.initializers.Ones(),
+        )(x)
+        model = keras.Model(inputs=x, outputs=y)
+
+        bentoml.tensorflow.save(model, "my_keras_model")
+
+:code:`bentoml.tensorflow` supports saving models with multiple tensors as input.
```
[the remark this answers is not in the document]
Comment (sauyon) [reply docs/source/frameworks/tensorflow.rst:100]:
```suggestion
``bentoml.tensorflow`` also supports saving models that take multiple tensors as input:
```

File: docs/source/frameworks/tensorflow.rst [hunk 13]
```diff
@@ -1,52 +1,244 @@
...
+            ...
+
+    model = MultiInputModel()
+    ... # training
+
+    bentoml.tensorflow.save(model, "my_tf_model")
+
+.. note::
+
+    :code:`bentoml.tensorflow.save_model` has two parameters :code:`tf_signature` and :code:`signatures`.
```
[the remark this answers is not in the document]
Comment (sauyon) [reply docs/source/frameworks/tensorflow.rst:126]:
```suggestion
    :code:`bentoml.tensorflow.save_model` has two parameters: ``tf_signature`` and ``signatures``.
```

**Review by sauyon — COMMENTED** [review]
Halfway through this, will circle back after meetings.

File: docs/source/frameworks/tensorflow.rst (hunk 6 above)
Comment (ssheng) [reply docs/source/frameworks/tensorflow.rst]: Discussed with @sauyon, we think Keras doc should have a section for recommending converting Keras model to TensorFlow for better performance. In this section of the TF doc, we can advise not to use Keras in production and link to the conversion section of the Keras doc.

File: docs/source/frameworks/tensorflow.rst (hunk 7 above)
Comment (ssheng) [reply docs/source/frameworks/tensorflow.rst:135]:
```suggestion
    Use the following arguments to define the model signatures signatures ensure consistent model behaviors in a Python session and from the BentoML model store.
    - `tf_signatures` is an alias to `tf.saved_model.save <https://www.tensorflow.org/api_docs/python/tf/saved_model/save>`_ *signatures* field. This optional signatures controls which methods in a given `obj <https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/trackable/base.py#L281>`_ will be available to programs that consume `SavedModel's <https://www.tensorflow.org/guide/saved_model>`_, for example, serving APIs. Read more about TensorFlow's signatures behavior `from their API documentation <https://www.tensorflow.org/api_docs/python/tf/saved_model/save>`_.
    - ``signatures`` refers to a general :ref:`Model Signatures <concepts/model:Model Signatures>`_ that dictates which methods can be used for inference in the Runner context. This signatures dictionary will be used during the creation process of a Runner instance.

The signatures used for creating a Runner is ``{"__call__": {"batchable": False}}``. This means by default, BentoML’s `Adaptive Batching <guides/batching:Adaptive Batching>`_ is disabled when using :obj:`~bentoml.tensorflow.save_model()`. If you want to utilize adaptive batching behavior and know your model's dynamic batching dimension, make sure to pass in ``signatures`` as follow: 

```

File: docs/source/frameworks/tensorflow.rst (hunk 6 above)
Comment (bojiang) [reply docs/source/frameworks/tensorflow.rst]: @sauyon We have

File: docs/source/frameworks/tensorflow.rst [hunk 14]
```diff
@@ -1,52 +1,290 @@
...
+    bentoml.tensorflow.save(
+        model,
+        "my_tf_model",
+        signatures={"__call__": {"batchable": True, "batchdim": 0}}
+    )
+
+.. note::
+
+    :obj:`~bentoml.tensorflow.save_model` has two parameters: ``tf_signature`` and ``signatures``.
+    Use the following arguments to define the model signatures signatures ensure consistent model behaviors in a Python session and from the BentoML model store.
```
Comment (ssheng) [inline docs/source/frameworks/tensorflow.rst:187]:
```suggestion
    Use the following arguments to define the model signatures to ensure consistent model behaviors in a Python session and from the BentoML model store.
```

File: docs/source/frameworks/tensorflow.rst [hunk 15]
```diff
@@ -1,52 +1,290 @@
...
+    Follow the steps to get the best performance out of your TensorFlow model.
+    #. Save the model with well-defined :obj:`~tf.function` decorator.
+    #. Apply adaptive batching if possible.
+    #. Serve on GPUs if applicable.
+    #. See performance guide from [TensorFlow Doc]
+
+Adaptive Batching
+-----------------
+
+Most TensorFlow models can accept batched data as input. If batched interence is supported, it is recommended to enable batching to take advantage of
```
Comment (ssheng) [inline docs/source/frameworks/tensorflow.rst:228]:
```suggestion
Most TensorFlow models can accept batched data as input. If batch inference is supported, it is recommended to enable batching to take advantage of 
```

**Review by ssheng — APPROVED** [review]

**Pull request merged**